File: src/word.cpp
#include <iostream>
#include <cstring>
#include "word.h"
using namespace std;

// Constructor Implementation
Word::Word(char* eng, char* chi){
    if(this->isEnglish(eng))
        this->setEnglish(eng);
    else{
        cout << "The first string should be in english..." << endl
             << "The input is: " << eng << endl
             << "set as 'None'..." << endl;
        this->setEnglish("None");
    }
    if(this->isChinese(chi))
        this->setChinese(chi);
    else{
        cout << "第二個字串須為中文 ..." << endl
             << "輸入為： " << chi << endl
             << "設為 '錯誤' ..." << endl;
        this->setEnglish("錯誤");
    }

    // Set default value
    Date d;
    d.Now();
    this->setDate(d);
    this->setNumberOfWrong(0);
    this->setNumberOfWatch(0);
}

/*
	Insert Method

	Arg: the value want to set
	Ret: If successfully set (Or none)
*/
void Word::setEnglish(const char* eng){
    strcpy(englist, eng);
}

void Word::setChinese(const char* chi){
    strcpy(chinese, chi);
}

void Word::setDate(Date d){
    date = d;
}

bool Word::setNumberOfWrong(int _){
    if(_>0){
        numberOfWrong = _;
        return true;
    }
    return false;
}

bool Word::setNumberOfWatch(int _){
    if(_>0){
        numberOfWrong = _;
        return true;
    }
    return false;
}

/*
	Get Method

	Ret: The corresponding value
*/

const char* Word::getEnglish() const{
    return englist;
}

const char* Word::getChinese() const{
    return chinese;
}

Date Word::getDate() const{
    return date;
}

int Word::getNumberOfWrong() const{
    return numberOfWrong;
}

int Word::getNumberOfWatch() const{
    return numberOfWatch;
}

// Other method
Word& Word::operator=(const Word &wrd) {
	this->setEnglish(wrd.getEnglish());
    this->setChinese(wrd.getChinese());
    this->setDate(wrd.getDate());
    this->setNumberOfWrong(wrd.getNumberOfWrong());
    this->setNumberOfWatch(wrd.getNumberOfWatch());
	return *this;  					// Return a reference to myself.
}

// ----- Judge method -----
/*
    Judge if the string is in english.
    It depend on the first two character.

    Arg: string
    Ret: if it's english'
*/
bool Word::isEnglish(const char* _){
    if(sizeof(_) < 2*sizeof(char)){
        cout << "The string is too short to judge" << endl;
        return false;
    }
    if( int(_[0])+int(_[0])>58 && 
        int(_[0])+int(_[0])<256 )
        return true;
    return false;
}

/*
    Judge if the string is in english.
    It's tricky that it would judge as chinese if it's not english

    Arg: string
    Ret: if it's english'
*/
bool Word::isChinese(const char* _){
    return !(this->isEnglish(_));
}

File: android/Ivy/app/src/main/java/edu/sunner/ivy/Word.java
package edu.sunner.ivy;

import android.util.Log;

/**
 * The Word structure
 *
 * @author sunner
 * @since 9/4/16.
 */
public class Word {
    // ----- Content -----
    private String english;
    private String chinese;
    private Date date;
    private int numberOfWrong;
    private int numberOfWatch;

    // ----- Method -----
    /**
     * Constructor.
     */
    public Word() {
        numberOfWatch = 0;
        numberOfWrong = 0;
        date = new Date();
        date.now();
    }

    public Word(String english, String chinese) {
        this.english = english;
        this.chinese = chinese;
    }

    // Insert method
    public void setEnglish(String eng) {
        english = eng;
    }

    public void setChinese(String chi) {
        chinese = chi;
    }

    public void setDate(Date date) {
        this.date = date;
    }

    /**
     * Set the number of wrong with the specific value.
     *
     * @param value the value want to set
     * @return if successfully sets
     */
    boolean setNumberOfWrong(int value) {
        if (value > 0) {
            numberOfWrong = value;
            return true;
        }
        return false;
    }

    /**
     * Set the number of watch with the specific value.
     *
     * @param value the value want to set
     * @return if successfully sets
     */
    boolean setNumberOfWatch(int value) {
        if (value > 0) {
            numberOfWatch = value;
            return true;
        }
        return false;
    }

    // Get method
    public String getEnglish() {
        return english;
    }

    public String getChinese() {
        return chinese;
    }

    public Date getDate() {
        return date;
    }

    public int getNumberOfWrong() {
        return numberOfWrong;
    }

    public int getNumberOfWatch() {
        return numberOfWatch;
    }

    /**
     * Judge if the string is in english.
     *
     * @param string the string want to examine
     * @return if it's in english
     */
    public boolean isEnglish(String string) {
        if (string.length() < 2) {
            Log.d(Constant.WD_TAG, "The string is too short to judge");
            return false;
        }
        if (Character.getNumericValue(string.charAt(0))
            + Character.getNumericValue(string.charAt(1)) > 58
            && Character.getNumericValue(string.charAt(0))
                + Character.getNumericValue(string.charAt(1)) < 256) {
            return true;
        }
        return false;
    }

    /**
     * Judge if the string is in english.
     * Assume there're only english and chinese two cases
     *
     * @param string the string want to examine
     * @return if it's in chinese
     */
    public boolean isChinese(String string) {
        return !isEnglish(string);
    }

    /**
     * Convert the Word object to string.
     *
     * @return the string
     */
    public String toString() {
        String res = "";
        res = res + english + "\n";
        res = res + chinese + "\n";
        res = res + date.toString() + "\n";
        res = res + numberOfWrong + "\n";
        res = res + numberOfWatch + "\n";
        return res;
    }

    /**
     * Show the Word object contain.
     */
    public void show() {
        Log.d(Constant.WD_TAG, "English: " + english + "\nChinese: " + chinese
            + "\nDate: " + date.toString() + "\nNumber of wrong: " + numberOfWrong
            + "\nNumber of watch: " + numberOfWatch);
    }
}
